### PR TITLE
Java deploy fix

### PR DESCRIPTION
**Pull request by SumiranAgg**

### CODE REVIEW

File: Tasks/Common/webdeployment-common/utility.ts [hunk 1]
```diff
@@ -183,6 +183,12 @@ export async function generateTemporaryFolderForDeployment(isFolderBasedDeployme
     if(isFolderBasedDeployment || packageType === PackageType.jar) {
         tl.debug('Copying Web Packge: ' + webDeployPkg + ' to temporary location: ' + folderPath);
         copyDirectory(webDeployPkg, folderPath);
+        if(packageType === PackageType.jar) {
+            let src = path.join(folderPath, getFileNameFromPath(webDeployPkg));
+            let dest = path.join(folderPath, "app.jar")
+            tl.debug("Renaming " + src + " to " + dest);
+            fs.rename(src, dest);
```
Comment (asranja) [inline Tasks/Common/webdeployment-common/utility.ts]: Shouldn't we use fs.[renameSync ](https://nodejs.org/api/fs.html#fs_fs_renamesync_oldpath_newpath)?

File: Tasks/Common/webdeployment-common/utility.ts [hunk 2]
```diff
@@ -183,6 +183,12 @@ export async function generateTemporaryFolderForDeployment(isFolderBasedDeployme
...
+        }
```
Comment (asranja) [inline Tasks/Common/webdeployment-common/utility.ts]: Leave a line after closing braces.

File: Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts [hunk 3]
```diff
@@ -182,6 +183,12 @@ export async function generateTemporaryFolderForDeployment(isFolderBasedDeployme
     if(isFolderBasedDeployment || packageType === PackageType.jar) {
         tl.debug('Copying Web Packge: ' + webDeployPkg + ' to temporary location: ' + folderPath);
         copyDirectory(webDeployPkg, folderPath);
+        if(packageType === PackageType.jar) {
+            let src = path.join(folderPath, getFileNameFromPath(webDeployPkg));
+            let dest = path.join(folderPath, "app.jar")
+            tl.debug("Renaming " + src + " to " + dest);
+            fs.rename(src, dest);
```
Comment (asranja) [inline Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts]: fs.renameSync ?

File: Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts [hunk 4]
```diff
@@ -3,6 +3,7 @@ import tl = require('vsts-task-lib/task');
 import { PackageType } from './packageUtility';
 import zipUtility = require('./ziputility.js');
 import * as os from "os";
+import fs = require('fs');
```
Comment (asranja) [inline Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts]: import * as fs from "fs"

File: Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts [hunk 5]
```diff
@@ -217,6 +224,10 @@ export class AzureAppServiceUtility {
         }
     }
 
+    private startupCommandRecommendation(runtimeStack: string) {
+        // Add startup command recommendations for different runime environments
```
Comment (asranja) [inline Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts]: Is this a TODO and beyond the scope of the PR ?

File: Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts [hunk 6]
```diff
@@ -4,6 +4,9 @@ import webClient = require('../azure-arm-rest/webClient');
 var parseString = require('xml2js').parseString;
 import Q = require('q');
 import { Kudu } from '../azure-arm-rest/azure-arm-app-service-kudu';
+import { Package, PackageType } from '../webdeployment-common/packageUtility';
+
+var webCommonUtility = require('azurermdeploycommon/webdeployment-common/utility.js');
```
Comment (asranja) [inline Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts]: Is this required ?

File: Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts [hunk 7]
```diff
@@ -202,13 +205,17 @@ export class AzureAppServiceUtility {
         }
     }
 
-    public async updateStartupCommandAndRuntimeStack(runtimeStack: string, startupCommand?: string): Promise<void> {
+    public async updateStartupCommandAndRuntimeStack(runtimeStack: string, startupCommand?: string, appPackage?: Package): Promise<void> {
```
Comment (asranja) [inline Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts]: Where are we using appPackage ?

File: Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts (hunk 5 above)
Comment (SumiranAgg) [reply Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts]: Yes. Its a place holder to add some recommendations as followup of this PR

File: Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts (hunk 7 above)
Comment (SumiranAgg) [reply Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts]: removed.

File: Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts (hunk 6 above)
Comment (SumiranAgg) [reply Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts]: removed

File: Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts (hunk 4 above)
Comment (SumiranAgg) [reply Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts]: changed

File: Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts (hunk 3 above)
Comment (SumiranAgg) [reply Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts]: changed

File: Tasks/Common/webdeployment-common/utility.ts (hunk 2 above)
Comment (SumiranAgg) [reply Tasks/Common/webdeployment-common/utility.ts]: done

File: Tasks/Common/webdeployment-common/utility.ts (hunk 1 above)
Comment (SumiranAgg) [reply Tasks/Common/webdeployment-common/utility.ts]: done

**Comment by Ajay-MS** [comment]
There is an L0 test which force to update all task versions which are being updated by this common module.

File: Tasks/Common/AzureRmDeploy-common/azure-arm-rest/azure-arm-app-service-kudu.ts [hunk 8]
```diff
@@ -455,7 +455,7 @@ export class Kudu {
         httpRequest.body = fs.createReadStream(webPackage);
 
         try {
-            let response = await this._client.beginRequest(httpRequest, null, 'application/octet-stream');
+            let response = await this._client.beginRequest(httpRequest, null, 'multipart/form-data');
```
Comment (kmkumaran) [inline Tasks/Common/AzureRmDeploy-common/azure-arm-rest/azure-arm-app-service-kudu.ts]: Why is this change required?

File: Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts [hunk 9]
```diff
@@ -182,6 +183,13 @@ export async function generateTemporaryFolderForDeployment(isFolderBasedDeployme
     if(isFolderBasedDeployment || packageType === PackageType.jar) {
         tl.debug('Copying Web Packge: ' + webDeployPkg + ' to temporary location: ' + folderPath);
         copyDirectory(webDeployPkg, folderPath);
+        if(packageType === PackageType.jar) {
+            let src = path.join(folderPath, getFileNameFromPath(webDeployPkg));
+            let dest = path.join(folderPath, "app.jar")
```
Comment (kmkumaran) [inline Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts]: I see we are taking the approach of renaming the jar file to app.jar. Does it impact the application. Typically the app name acts as virtual path (ex: webpp.azurewebsites.net/<appName>). With renaming it to default app.jar, I presume there wouldn't be no virtual path once the app deploys. Is my understanding right?

File: Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts (hunk 5 above)
Comment (kmkumaran) [reply Tasks/Common/AzureRmDeploy-common/operations/AzureAppServiceUtility.ts]: Let's add this when required.

File: Tasks/Common/AzureRmDeploy-common/azure-arm-rest/azure-arm-app-service-kudu.ts (hunk 8 above)
Comment (SumiranAgg) [reply Tasks/Common/AzureRmDeploy-common/azure-arm-rest/azure-arm-app-service-kudu.ts]: According to this document https://docs.microsoft.com/en-us/azure/app-service/deploy-zip multipart/form is the recommended content-type for both zip and war deploy.

File: Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts (hunk 9 above)
Comment (SumiranAgg) [reply Tasks/Common/AzureRmDeploy-common/webdeployment-common/utility.ts]: As per my understanding, package name cannot act as virtual path, since package is not unique. This is one of the recommendations by app service team. Also azure maven plugin team is doing the same.

File: Tasks/Common/AzureRmDeploy-common/azure-arm-rest/azure-arm-app-service-kudu.ts (hunk 8 above)
Comment (kmkumaran) [reply Tasks/Common/AzureRmDeploy-common/azure-arm-rest/azure-arm-app-service-kudu.ts]: @vincentdass - I recollect you had changed the content type earlier to octet-stream for fixing an issue. Can you review this to see we are not regressing anything now.

File: Tasks/AzureWebAppV1/deploymentProvider/BuiltInLinuxWebAppDeploymentProvider.ts [hunk 10]
```diff
@@ -32,30 +34,18 @@ export class BuiltInLinuxWebAppDeploymentProvider extends AzureRmWebAppDeploymen
...
-                    this.taskParams.AppSettings += " -INIT_SCRIPT " + destInitScriptPath;
-                }
-                this.taskParams.AppSettings = this.taskParams.AppSettings.trim();
-                tl.writeFile(initScriptFile, script, { encoding: 'utf8' });
-                var output = await webCommonUtility.archiveFolderForDeployment(false, folderPath);
-                var webPackage = output.webDeployPkg;
+                let folderPath = await webCommonUtility.generateTemporaryFolderForDeployment(false, this.taskParams.Package.getPath(), PackageType.jar);
+                let output = await webCommonUtility.archiveFolderForDeployment(false, folderPath);
+                let webPackage = output.webDeployPkg;
+                let deleteCustomApplicationSetting = ParameterParser.parse(initScriptAppSetting);
```
[the remark this answers is not in the document]
Comment (SumiranAgg) [reply Tasks/AzureWebAppV1/deploymentProvider/BuiltInLinuxWebAppDeploymentProvider.ts]: Yes
